### ISSUE TITLE
Live CD: Investigate a custom Apparmor profile

### ISSUE THREAD

**Issue by yveszoundi** [rimerosolutions/entrusted#18]
## Background

There are couple of hardening options implemented on the Live CD starting with 0.3.0

- Kernel settings ([Grub boot settings and sysctl.conf](https://madaidans-insecurities.github.io/guides/linux-hardening.html))
- Web interface started with [hardened malloc](https://github.com/GrapheneOS/hardened_malloc/) as memory allocator
- [Minimal systemd unit sandboxing](https://www.digitalocean.com/community/tutorials/how-to-sandbox-processes-with-systemd-on-ubuntu-20-04) options that still allows the Web interface to run correctly
- [container seccomp security profile](https://docs.docker.com/engine/security/seccomp/), just like in the CLI and Desktop interfaces
- [Few SSH hardening options](https://node-security.com/posts/ssh-server-hardening/)

It would be nice to build on top of previous efforts with [AppArmor](https://apparmor.net/), especially for the Live CD which is a fully controlled environment. A [custom AppArmor security profile](https://docs.docker.com/engine/security/apparmor/) could be loaded during file processing.

## Objectives

- [x] Assess complexity of defining and loading an `AppArmor` profile
- [x] Assess ISO image resulting size after adding `AppArmor`
- [x] Assess CI/CD pipeline changes (**most of the work** is here)
- [x] Assess additional required changes in the `entrusted-client` component
  - We need the ability to pass additional `security-opt` flags to the container solution
  - Those additional flags would need to be a configurable setting (environment variable maybe??)

**Comment by yveszoundi** [comment]
tl;dr this should work with [Nerdctl](https://github.com/containerd/nerdctl/) instead of [podman](https://github.com/containers/podman) on the Live CD, 
- `nerdctl` can operate as rootless once it's aware of available `apparmor` profiles.
-  _This does work locally_ and just needs to be "ported" to the Live CD properly...

### Appamor profile creation

This is not that complicated and the `apparmor` tooling helps fine-tune the expected permissions.

### ISO image size 
 
- The size of the ISO image seems to have grown from `~800 MB` to `~1.4 GB`
  - Is it because of snapshots??
  - Is it because of the storage format?? 

### CI/CD pipeline

#### Replacing Podman with Nerdctl

- Download the latest release of _nerdctl_ during the build process
- Similar to _podman_ prepare _nerdctl_ and make sure that it's setup as _rootless_
  - Set `XDG_DATA_DIR` and `PATH` (including `nerdctl`) in the `.bash_profile` of the user
  - Enable `lingering` for the local `entrusted` user
  - Start `dbus`, as the local `entrusted` user
  - Start `containerd`, as the local `entrusted` user
- Instead of previously loading a tar container image via _podman_, do it with _nerdctl_
- Install few more packages on the Live CD: `rootlesskit`, `iproute2` and `containerd`
- Copy the relevant container storage files on the CD (`/home/entrusted/.local` folders)
- Copy the `nerdctl` binaries to the Live CD prior the ISO image generation

#### Enabling Apparmor on boot

- In `grub.cfg` and `isolinux.cfg`, add new boot parameters: `apparmor=1 security=apparmor`
- Ensure that _apparmor_ packages are installed on the Live CD: `apparmor` and `apparmor-utils`
- Enable _apparmor_ on startup `systemctl enable apparmor`
- Ensure that _nerdctl_ has _apparmor_ profiles loaded
  - `apparmor_parser -r -W /etc/apparmor.d/usr/local.bin.entrusted-container` 
  - `nerdctl apparmor load`
 
### Application changes

A new build `feature` (`cargo` terminology) could help inject pre-defined settings for the Live CD

- That wouldn't be a desired approach for command-line parameters editing (`security-opt` flag)
- Enabling `gVisor` would require similar tricks (#17 )
- Sadly we can't preset the apparmor profile without fatal errors
  - The `apparmor` needs to be loaded from the host system and must exists
  - The container solution must be "aware" of loaded apparmor profiles, otherwise conversions will just fail almost immediately from a user perspective (container process failure to start...)

**Comment by yveszoundi** [comment]
An unused `apparmor` profile will be checked-in in the main tree of the repository, so that previous efforts don't disappear locally (or in an old stale GitHub branch later). Most of what needs to happen is described in previous comments.

### Local machine results

Prior running the command the `safezone` folder is marked as read-write for all `chmod  -R a+rw`.

```
$ sudo nerdctl apparmor load
$ nerdctl run --network none --cap-drop all --security-opt no-new-privileges --security-opt apparmor=/usr/local/bin/entrusted-container --security-opt seccomp=/tmp/entrusted/seccomp-entrusted-profile-0.3.1.json -v /home/vip/test_docs/sample.odt:/tmp/input_file:Z -v /home/vip/testing/tmp/safezone:/safezone:Z -e ENTRUSTED_LOG_FORMAT=json -e ENTRUSTED_LANGID=en -e VERBOSE=1 -e ENTRUSTED_VISUAL_QUALITY=medium docker.io/uycyjnzgntrn/entrusted_container:0.3.1 /usr/local/bin/entrusted-container

WARN[0000] unsupported volume option "Z"
WARN[0000] unsupported volume option "Z"
{"percent_complete":5,"data":"Converting to PDF using LibreOffice"}
{"percent_complete":20,"data":"Extract PDF file into 5 PNG images"}
{"percent_complete":25,"data":"Extracting page 1 into a PNG image"}
{"percent_complete":30,"data":"Extracting page 2 into a PNG image"}
{"percent_complete":35,"data":"Extracting page 3 into a PNG image"}
{"percent_complete":40,"data":"Extracting page 4 into a PNG image"}
{"percent_complete":45,"data":"Extracting page 5 into a PNG image"}
{"percent_complete":45,"data":"Saving 5 PNG images to PDF"}
{"percent_complete":54,"data":"Saving PNG image 1 to PDF"}
{"percent_complete":63,"data":"Saving PNG image 2 to PDF"}
{"percent_complete":72,"data":"Saving PNG image 3 to PDF"}
{"percent_complete":81,"data":"Saving PNG image 4 to PDF"}
{"percent_complete":90,"data":"Saving PNG image 5 to PDF"}
{"percent_complete":90,"data":"Combining 5 PDF documents"}
{"percent_complete":91,"data":"Collecting PDF pages"}
{"percent_complete":92,"data":"Updating bookmarks and page numbering"}
{"percent_complete":93,"data":"Processing PDF structure"}
{"percent_complete":94,"data":"Updating PDF dictionnary"}
{"percent_complete":95,"data":"Combining PDF objects"}
{"percent_complete":96,"data":"Compressing PDF"}
{"percent_complete":98,"data":"Saving PDF"}
{"percent_complete":98,"data":"Moving output files to their final destination"}
{"percent_complete":99,"data":"Conversion succeeded!"}
{"percent_complete":100,"data":"Elapsed time: 0 hours 0 minutes 1 second"}
```

### Challenges on the Live CD

Not sure exactly what the permission issues are at this time. It's probably worth [talking with the nerdctl team](https://github.com/containerd/nerdctl/discussions) in the future.

```
entrusted@entrusted-livecd:~$ /usr/local/bin/nerdctl run --rm --network none --cap-drop all --security-opt no-new-privileges --security-opt seccomp=/tmp/entrusted/seccomp-entrusted-profile-0.3.1.json -v /home/entrusted/hello.txt:/tmp/input_file:Z -v /home/entrusted/safezone:/safezone:Z -e ENTRUSTED_LOG_FORMAT=json -e ENTRUSTED_LANGID=en -e ENTRUSTED_VISUAL_QUALITY=medium docker.io/uycyjnzgntrn/entrusted_container:0.3.1 /usr/local/bin/entrusted-container
WARN[0000] unsupported volume option "Z"                
WARN[0000] unsupported volume option "Z"                
FATA[0000] failed to mount /run/user/1024/containerd-mount3578917145: operation not permitted 
```

**Comment by yveszoundi** [comment]
Just like for `gVisor`, `apparmor` support will be revisited in the future: solving permission issues might not be that straightforward...
